task/mgmt-gateway/src/mgs_handler.rs

```diff
@@ -152,4 +221,104 @@ impl SpHandler for MgsHandler {
             Err(ResponseError::Busy)
         }
     }
+
+    fn sys_reset_prepare(
+        &mut self,
+        _sender: SocketAddrV6,
+        _port: SpPort,
+    ) -> Result<(), ResponseError> {
+        ringbuf_entry!(Log::MgsMessage(MgsMessage::SysResetPrepare));
+        self.reset_requested = true;
+        Ok(())
+    }
+
+    fn sys_reset_trigger(
+        &mut self,
+        _sender: SocketAddrV6,
+        _port: SpPort,
+    ) -> Result<Infallible, ResponseError> {
+        if !self.reset_requested {
+            return Err(ResponseError::SysResetTriggerWithoutPrepare);
+        }
+
+        let jefe = task_jefe_api::Jefe::from(crate::JEFE.get_task_id());
+        jefe.request_reset();
+
+        // If `request_reset()` returns, something has gone very wrong.
+        panic!()
+    }
+}
+
+#[derive(Default)]
+struct UpdateBuffer {
+    total_length: usize,
+    bytes_written: usize,
+    current_block: ArrayVec<[u8; BLOCK_SIZE_BYTES]>,
+}
+
+impl UpdateBuffer {
+    fn ingest_chunk(
+        &mut self,
+        update_task: &Update,
+        chunk: &UpdateChunk,
+    ) -> Result<(), ResponseError> {
+        // Reject chunks that don't match our current progress.
+        if chunk.offset as usize != self.bytes_written {
+            return Err(ResponseError::UpdateInProgress {
+                bytes_received: self.bytes_written as u32,
+            });
+        }
+
+        // Reject chunks that would go past the total size we're expecting.
+        if self.bytes_written + chunk.chunk_length as usize > self.total_length
+        {
+            return Err(ResponseError::InvalidUpdateChunk);
+        }
+
+        let mut data = &chunk.data[..chunk.chunk_length as usize];
+        while !data.is_empty() {
+            let cap = self.current_block.capacity() - self.current_block.len();
+            assert!(cap > 0);
+            let to_copy = usize::min(cap, data.len());
+
+            let current_block_index = self.bytes_written / BLOCK_SIZE_BYTES;
+            self.current_block.extend_from_slice(&data[..to_copy]);
+            data = &data[to_copy..];
+            self.bytes_written += to_copy;
+
+            // If the block is full or this is the final block, send it to the
+            // update task.
+            if self.current_block.len() == self.current_block.capacity()
+                || self.bytes_written == self.total_length
+            {
+                update_task
+                    .write_one_block(current_block_index, &self.current_block)
+                    .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
```

Comment:
If this returns with an error, then we exit before clearing `self.current_block`, meaning the assertion at line 281 could fire if this function is called again.  Should we be robust to that sequence of events?

Reply:
Nice catch! Yes, we should, although as @kc8apf pointed out, failure mid update is currently a bit precarious. I'll update this to locally leave `UpdateBuffer` in a more consistent state so at least it can resume without panicking if given reasonable inputs.